File: go/okapi/DidKey.go
package okapi

import (
	okapi "github.com/trinsic-id/okapi/go/okapi/proto"
)

// DidKeyer is the interface that groups the did:key functions
type DidKeyer interface {
	Generate(request *okapi.GenerateKeyRequest) (*okapi.GenerateKeyResponse, error)
	Resolve(request *okapi.ResolveRequest) (*okapi.ResolveResponse, error)
}

// DidKey returns a DidKeyer that can generate and resolve did:keys
func DidKey() DidKeyer {
	return &didKey{}
}

type didKey struct{}

func (d *didKey) Generate(request *okapi.GenerateKeyRequest) (*okapi.GenerateKeyResponse, error) {
	response := okapi.GenerateKeyResponse{}
	err := callOkapiNative(request, &response, didkeyGenerate)
	return &response, err
}

func (d *didKey) Resolve(request *okapi.ResolveRequest) (*okapi.ResolveResponse, error) {
	response := okapi.ResolveResponse{}
	err := callOkapiNative(request, &response, didkeyResolve)
	return &response, err
}

File: ruby/lib/okapi/did_key.rb
# frozen_string_literal: true

module Okapi
  # DidKey module
  module DidKey
    def self.generate(request)
      Okapi.verify_type(request, Okapi::Keys::V1::GenerateKeyRequest)
      Okapi.ffi_call('didkey_generate', request, Okapi::Keys::V1::GenerateKeyResponse)
    end

    def self.resolve(request)
      Okapi.verify_type(request, Okapi::Keys::V1::ResolveRequest)
      Okapi.ffi_call('didkey_resolve', request, Okapi::Keys::V1::ResolveResponse)
    end
  end
end
